doc/Architecture.md

```diff
@@ -0,0 +1,53 @@
+
+Both the frontend and backend processes have their dependency injection container (see below) to which extensions can contribute.
+
+## Frontend
+
+The frontend process represents the client and renders the UI. In the browser, it simply runs in the rendering loop, while in Electron it runs in an Electron Window, which basically is a browser with additional Electron and node.js APIs. Therefore, frontend code may only access sources that depend on any API that can run in browsers.
+
+The startup of the frontend process will first load the modules of all contributing extensions before it obtains an instance of `FrontendApplication` from the resulting Dependency Injection (DI) container and call `start()` on it.
+
+## Backend
+
+ The backend process runs on Node.js. We use express as the HTTP server. The backend also serves the static code for the frontend. So the backend is started first and exposes a port to serve the frontend application to a requesting browser. For Electron we open an `Electron.BrowserWindow` which is a chromium with some additional Electron APIs.
+
+ The startup of the backend application will first load the modules of all contributing extensions before it obtains an instance of `BackendApplication` from the resulting DI container and calls `start(portNumber)` on it.
+
+## Separation By Platform
+
+In an extension's top folder we have an additional layer of folders to separate by platform:
+
+ - The `common` folder contains code that doesn't depend on any runtime.
+ - The `browser` folder contains code requiring a modern browser as a platform (DOM API).
+ - The `electron-browser` folder contains frontend code that requires DOM API as well as Electron renderer-process specific APIs
+ - The `node` folder contains (backend) code requiring Node.js as a platform.
+ - The `node-electron` folder contains (backend) code specific for Electron.
+
+## Dependency Injection (DI)
+
+Theia uses the DI framework [Inversify.js](http://inversify.io/) to wire up the different components. Using DI decouples a certain component from creating its dependencies. Instead, it gets them injected in when created (as parameters of a constructor). A DI container does the creation for you, based on some configuration you provide on startup through so-called modules. For instance, the `Navigator` widget needs access to a `FileSystem` to present folders and files in a tree. The `FileSystem` itself is just a proxy sending JSON-RPC messages to some backend, so it needs a particular configuration and treatment. The navigator doesn't need to care as it will get injected a fully working `FileSystem` instance. Also in case, we want to change the actual implementation no one needs to change any code in the navigator. Moreover, this decoupling of construction and use, allows extensions to provide their very specific implementations of e.g. a `FileSystem` if needed. Still without touching any users of the `FileSystem`interface.
+
```

Comment:
Rewording maybe:
>Using DI decouples a certain component from creating its dependencies.
Using DI decouples components from it's dependencies ?
Or Using DI components are not responsible for the creation of their dependencies

 For instance, the `Navigator` widget needs access to a `FileSystem` to present folders and files in a tree.  With DI the concretion of that `FileSystem` interface it not important to the `Navigator` widget.  It can safely assume that an object consistent with the `FIleSystem` interface it ready to use. 

For example: In Theia the used `FileSystem` concretion is just a proxy sending JSON-RPC messages to some backend, so it needs a particular configuration and treatment. The navigator doesn't need to care as it will get injected a fully working `FileSystem` instance.

>Moreover, this decoupling of construction and use,
I would change construction and use with: interface and concretion ?  May need to adapt the rest of the paragraph as such ?

I would be more clear since it follows the wording someone will read at inversify.io